### ISSUE TITLE
Provide gcroot API

### ISSUE THREAD

**Issue by ivmai** [ivmai/bdwgc#378]
Probably it would be convenient to have API as in https://github.com/uim/libgcroots/blob/master/include/gcroots.h (to be able to write your own garbage collector w/o arch-specific code).

**Comment by ivmai** [comment]
relates #513

**Comment by akofer** [comment]
have a look at libgcroots. still lack of thread support, not usable in production.

**Comment by ivmai** [comment]
> have a look at libgcroots. still lack of thread support, not usable in production.

Yes, it is just an idea to implement something similar - a facility to get roots if someone wants to implement a garbage collector.

**Comment by ivmai** [comment]
Some notes about libgcroots (for me):
* updated LICENSE README AUTHORS (3 contributors)
* gcroots.pc.in (new file)
* the fork is based on bdwgc tag gc7_0
* configure: --disable-threads is default (because no threads support exists)
* include/gcroots.h (new file) GCROOTS_...
* libgcroots_la_...
* Change prefix to GCROOTS_: GC_push_regs, GC_push_one
* gcroots.c (new file)
* include: no changes in .h files, only gcroots.h and private/gcroots_priv.h are added
* mach_dep.c: almost no changes except for HAVE_GETCONTEXT checking
* mark.c: remove all functions but GC_noop1()
* loongarch support: https://github.com/uim/libgcroots/pull/6 (already in upstream)

**Comment by ivmai** [comment]
Hello @kou,
FYI: I'm considering to import changes/API of libgcroots (which is a fork of [libgc-7.0](https://github.com/bdwgc/bdwgc/releases/tag/gc7_0) to back to bdwgc. So that `libgcroot.so` to have latest targets support and fixes (as bdwgc has).

**Comment by kou** [comment]
Thanks for the ping!

If it's implemented, SigScheme will use bdwgc directly and we will deprecate libgcroots!

**Comment by kou** [comment]
Note that you don't need to use the same name as libgcroots for API. If the same functionalities are provided, SigScheme can adapt them.

**Comment by ivmai** [comment]
Here's the proposed solution:
1. Add `GC_custom_push_reg_and_stack` API function to bdwgc (which does not need `GC_init()` call)
2. Rewrite src/gcroots/gcroots.c (in [SigScheme](https://github.com/uim/sigscheme)) to use `GC_custom_push_reg_and_stack` and `GC_call_with_stack_base` instead of platform-specific code if `SCM_WITH_BDWGC` macro is defined
3. Add bdwgc (as git submodule) to SigScheme source and add support of `--with_libgcroots=bundled-bdwgc` option to configure (which configures bdwgc and defines `SCM_WITH_BDWGC` macro)

The drawback of this solution, however:
- client code has dependency on libgc.so
- yet another git submodule (later after making release of bdwgc with the added GC_custom_push_reg_and_stack it would be possible to add an option to use system-provided bdwgc instead of the bundled one).

In short, the solution looks overkilling :( but it makes implementation of src/gcroots platform-independent (except for `ALIGNOF_VOID_P` macro).

**Comment by ivmai** [comment]
I will provide the patch for SigScheme soon.

**Comment by ivmai** [comment]
I've added a PR (https://github.com/uim/sigscheme/pull/29) to SigScheme for review and further testing.
Though I'm sure if it is worth to merge it, at least because of adding bdwgc submodule (is there a way not to add it?).

**Comment by ivmai** [comment]
> have a look at [libgcroots](https://github.com/uim/libgcroots). still lack of thread support, not usable in production

Note: `GC_custom_push_reg_and_stack` and `GC_call_with_stack_base` are usable in the multi-threaded code (and coroutines), and supported on all platforms where libgc itself is supported. And the code in src/gcroots (with defined `SCM_WITH_BDWGC`) does not rely on global variables (i.e. it is usable in the multi-threaded code).

**Comment by ivmai** [comment]
Alternatively (or in addition to), we could update [libgcroots](https://github.com/uim/libgcroots) to fully rely on libgc, i.e. make libgcroots to be a tiny portable wrapper over libgc. But later, after, the new version of libgc (with `GC_custom_push_reg_and_stack`) reaches major OS distributions.

**Comment by kou** [comment]
Thanks!

We don't need to maintain libgcroots. SigScheme (and uim) will switch to bdwgc entirely and deprecate libgcroots eventually.
